### ISSUE TITLE
Generic with ParamSpec and Concatenate substitutions behave incorrectly

### ISSUE THREAD

**Issue by Daraan** [python/typing_extensions#492]
The following code displays some incorrect behavior, but different parts do not work in certain ranges of 3.8-3.11.3

- 3.8 & 3.9 substitution of `Concatenate`
- 3.10 does not correctly collect type parameters
- < 3.11.3 wrong substitution behavior; type parameter not removed

```python
from typing import Generic, TypeVar
from typing_extensions import ParamSpec, Concatenate
P = ParamSpec('P')
T = TypeVar("T")

class CLS(Generic[P]): ...

Y = CLS[T] # or with a ParamSpec
# 3.10
Y[str]  # raises TypeError is not a generic class

# 3.8 & 3.9
G9 = CLS[Concatenate[T, P]]
H9 = G9[int, [T]] # should be ok; but TypeError
H9[str]  # should also work

# 3.11.0
G8 = CLS[Concatenate[T, ...]]
H1 = G8[int]
H1[str]  # for python 3.11.0-3 this still has a parameter and does not raise a TypeError
```

This is a variant of #126 but with generics which require different fixes. There are multiple flavors of this issue depending on the version parameters are not recognized or not substituted correctly.

For a more comprehensive list check the skipped tests of #491 [here](https://github.com/python/typing_extensions/pull/491/files#diff-5548e20fd249808373a3a5176f402603ac2230e02e681c172dd528fe44e60638)

---

It's possible that some substitutions with `Concatenate` worked prior to #479 and caused a regression here, but I am not yet certain.

---

Fixes come with

- #489

**Comment by jeertmans** [comment]
No sure if this is directly related, but Python 3.11 added support for ellipsis `...` as last parameter in `Concatenate`. In Python 3.10, this was not possible (only `ParamSpec` was allowed).

It seems that `typing_extensions` provides a 3.10-compatible version of `Concatenate` when `Python < 3.11`, which means I cannot use `Concatenate[Foo, ...]` unless my library's minimal supported Python version is 3.11.

In other words, I was planning to do that:

```python
if sys.version_info >= (3, 11):
    from typing import Concatenate
else:
    from typing_extension import Concatenate

Bar = Concatenate[Foo, ...]
```

but this doesn't work

**Comment by Daraan** [comment]
> Not sure if this is directly related, but Python 3.11 added support for ellipsis `...` as last parameter in `Concatenate`. In Python 3.10, this was not possible (only `ParamSpec` was allowed).

The issue in question for this problem is #110 to address this problem for 3.10. A few weeks ago we added #481 to `main` which solves this problem, but note that there is no release yet.

**Comment by jeertmans** [comment]
Oh, thanks for pointing to the correct issue :-)

> > Not sure if this is directly related, but Python 3.11 added support for ellipsis `...` as last parameter in `Concatenate`. In Python 3.10, this was not possible (only `ParamSpec` was allowed).
> 
> The issue in question for this problem is #110 to address this problem for 3.10. A few weeks ago we added #481 to `main` which solves this problem, but note that there is no release yet.
